helix-view/src/tree.rs

```diff
@@ -521,6 +521,24 @@ impl Tree {
         Some(child_id)
     }
 
+    pub fn focus_prev(&mut self) {
+        let pos = self.traverse().position(|(id, _view)| id == self.focus);
+
+        if let Some(pos) = pos {
+            if pos > 0 {
+                match self.traverse().nth(pos - 1) {
+                    Some((id, _view)) => self.focus = id,
+                    None => {
+                        let (key, _) = self.traverse().next().unwrap();
+                        self.focus = key;
+                    }
+                }
+            } else {
+                self.focus_next();
+            }
+        }
+    }
```

Comment:
You could implement DoubleEndedIterator for Traverse and then this function would be almost the same as `focus_next`

Reply:
i.e. it should be possible to write `focus_prev` like this:

```rust
let mut views = self
    .traverse()
    .rev()
    .skip_while(|&(id, _view)| id != self.focus)
    .skip(1);
if let Some((id, _)) = views.next() {
    self.focus = id;
} else {
    let (key, _) = self.traverse().rev().next().unwrap();
    self.focus = key;
}
```